pkg/storj/metainfo.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storj
+
+import (
+	"context"
+)
+
+// Metainfo represents a database for storing meta-info about objects
+type Metainfo interface {
+	// MetainfoLimits returns limits for this metainfo database
+	Limits() (MetainfoLimits, error)
+
+	// CreateBucket creates a new bucket with the specified information
+	CreateBucket(ctx context.Context, bucket Bucket) (Bucket, error)
+	// DeleteBucket deletes bucket
+	DeleteBucket(ctx context.Context, bucket string) error
+	// GetBucket gets bucket information
+	GetBucket(ctx context.Context, bucket string) (Bucket, error)
+
+	// GetObject returns information about an object
+	GetObject(ctx context.Context, bucket string, path Path) (Object, error)
+	// GetObjectStream returns interface for reading the object stream
+	GetObjectStream(ctx context.Context, bucket string, path Path) (ReadonlyStream, error)
+
+	// CreateObject creates an uploading object and returns an interface for uploading Object information
+	CreateObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// ModifyObject creates an interface for modifying an existing object
+	ModifyObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// DeleteObject deletes an object from database
+	DeleteObject(ctx context.Context, bucket string, path Path) error
+
+	// TODO: add things for continuing existing objects
+	ListObjects(cctx context.Context, bucket string, options ListOptions) (ObjectList, error)
+}
+
+// ListOptions lists objects
+type ListOptions struct {
+	Prefix    Path
+	First     Path // First is relative to Prefix, full path is Prefix + First
```

Comment:
I like `StartAfter` better as it communicates clearer that it won't be included in the result. Or am I wrong about the semantics here? I see we have also the `Token`...

Reply:
I want the `First` to be included, I think it makes more sense API wise (i.e. otherwise there's no easy way to get the listing starting at a specific item).

The `Token` is there to simulate the S3 ListV2 behavior, but I'm not quite sure whether we want that to be specified. I understand that it's there so you could much better continue listing on the backend side without having to worry about the caller to specify the First/Prefix correctly from previous listing.

Reply:
With regards to `StartAfter` it is easy to simulate with `first + "\x00"`.

Reply:
How would the code look like for a caller who wants to list all the objects in a bucket?

Reply:
`ListObjects(ctx, "bucket", ListOptions{Recursive: true})`

Reply:
Thanks. This `NextFirst` field looks interesting. I don't see it defined in the `ObjectList` type.

Also, the above example should check the `objects.More` field before making another `ListObjects` call, shouldn't they?

Reply:
Yes, added the `objects.More` check.